server/src/main/scala/org/http4s/server/middleware/VirtualHost.scala

```diff
@@ -0,0 +1,41 @@
+package org.http4s
+package server
+package middleware
+
+import headers.Host
+import Status.BadRequest
+
+/** Middleware for virtual host mapping
+  *
+  * The `VirtualHost` middleware allows multiple services to be mapped
+  * based on the [[Host]] header of the [[Request]]. A default service
+  * is specified which will handle cases where the [[Host]] header is
+  * not present (should only happen for HTTP/1.0 requests) or if none
+  * of the hosts match.
+  * If the port is not specified by the client _or_ the server, it is
+  * ignored because the middleware cannot know the default port for the
+  * underlying transport.
+  */
+object VirtualHost {
+
+  /** Specification of the virtual host */
+  case class HostService(service: HttpService, host: String, port: Option[Int] = None)
+
+  def apply(default: HostService, rest: HostService*): HttpService = {
+
+    def check(host: Host, hs: String, port: Option[Int]): Boolean =
+      host.host == hs && (port.isEmpty || host.port.isEmpty || host.port.get == port.get)
+
+    val all = (default +: rest).toVector
+
+    Service.lift { req =>
+      req.headers.get(Host) match {
+        case None if req.httpVersion == HttpVersion.`HTTP/1.0` => default.service(req)
+        case None => Response(BadRequest).withBody("Missing Host header.")
+        case Some(host) =>
+          all.collectFirst { case HostService(s,h,p) if check(host,h,p) => s(req) }
+             .getOrElse(default.service(req))
```

Comment:
That behavior begs the question about what to do with http/1.0 requests. Would they result in an `505 HTTP Version Not Supported`?